src/js/letters/containers/AddressSection.jsx

```diff
@@ -43,38 +63,122 @@ export class AddressSection extends React.Component {
+   *                                         saveAddress will need to validate all
+   *                                         fields regardless of whether they've been
+   *                                         modified.
+   * @return {String|undefined}             If there's a validation error, return the
+   *                                         error message. If not, return undefined.
+   */
+  validateField = (fieldName, fullAddress, ignoreShouldValidate = false) => {
+    // Only validate the field if it's been modified
+    if (!this.state.shouldValidate[fieldName] || ignoreShouldValidate) {
+      return undefined;
+    }
+
+    const validations = fieldValidations[fieldName];
+    // If there is no validations array for that field, assume it has no validations
+    if (!validations) {
+      return undefined;
+    }
+
+    let errorMessage = false;
+    for (let i = 0; i < validations.length; i++) {
+      // this.props.value = address
+      errorMessage = validations[i](fullAddress[fieldName], fullAddress);
+      if (typeof errorMessage === 'string') {
+        return errorMessage;
+      }
+    }
+
+    // All validations passed; there are no error messages to report
+    return undefined;
+  }
+
+  /**
+   * Runs validation for all fields, returning a complete errorMessages object.
+   *
+   * @return {Object}  Holds all the error messages for all the fields that have them.
+   */
+  validateAll = (address = this.state.editableAddress) => {
+    const errorMessages = {};
+    Object.keys(fieldValidations).forEach((fieldName) => {
+      errorMessages[fieldName] = this.validateField(fieldName, address);
+    });
+
+    return errorMessages;
+  }
+
+  // TODO: Make sure this doesn't allow us to save the address if there are validation errors
+  saveAddress = () => {
+    const errorMessages = this.validateAll();
+    // If there are errors, show them, but don't stop editing and don't save
+    if (Object.keys(errorMessages).length === 0) {
+      this.setState({ errorMessages });
+      return;
+    }
 
+    this.setState({
+      isEditingAddress: false,
+      // Reset all the error messages in case they go to edit again; should be pointless
+      errorMessages: {}
+    });
     this.props.saveAddress(this.state.editableAddress);
   }
 
-  handleChange = (path, update) => {
-    this.setState(({ editableAddress }) => {
-      // reset state code when user updates country
-      const newFragment = (path === 'country'
-        ? {
-          [path]: update,
-          state: '',
-          militaryStateCode: '',
-        }
-        : { [path]: update }
-      );
+  /**
+   * Infers the address type from the address supplied and returns the address
+   *  with the "new" type.
+   */
+  inferAddressType = (address) => {
+    let type = ADDRESS_TYPES.domestic;
+    if (!['USA', 'US'].includes(address.country)) {
+      type = ADDRESS_TYPES.international;
+    } else if (address.militaryStateCode) {
```

Comment:
Hm, will this ever be true? Aren't we deleting `militaryStateCode` in the actions? Think we may need to instead check if the `['AE', 'AA', 'AP'].includes(address.state)` or something?

Reply:
This smells like a bad merge...Good catch; I'll look into it.